File: lib/services/character_group_community_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/character_group_package.dart';
import '../utils/api_config.dart';
import 'cloud_account_service.dart';

class CharacterGroupCommunityService {
  const CharacterGroupCommunityService();

  Future<List<CommunityCharacterGroupInfo>> listGroups({
    String? query,
    int limit = 40,
    int offset = 0,
  }) async {
    final json = await _requestJson(
      'GET',
      '/api/community/character-groups',
      queryParameters: {
        if (query != null && query.trim().isNotEmpty) 'keyword': query.trim(),
        'limit': limit.toString(),
        'offset': offset.toString(),
      },
    );
    final data = json['data'];
    final list = data is Map ? data['items'] : data;
    if (list is! List) return const [];
    return list
        .whereType<Map>()
        .map(
          (item) => CommunityCharacterGroupInfo.fromJson(
            Map<String, dynamic>.from(item),
          ),
        )
        .toList(growable: false);
  }

  Future<CharacterGroupPackage> fetchGroup(String id) async {
    final json = await _requestJson(
      'GET',
      '/api/community/character-groups/${Uri.encodeComponent(id)}',
    );
    final data = json['data'];
    if (data is Map) {
      return CharacterGroupPackage.fromJson(Map<String, dynamic>.from(data));
    }
    throw const CloudApiException('社区角色群组数据格式不正确');
  }

  Future<CharacterGroupPackage> fetchByShareCode(String code) async {
    final cleanCode = code.trim();
    if (cleanCode.isEmpty) throw const CloudApiException('请输入分享码');
    final json = await _requestJson(
      'GET',
      '/api/community/character-groups/share/${Uri.encodeComponent(cleanCode)}',
    );
    final data = json['data'];
    if (data is Map) {
      return CharacterGroupPackage.fromJson(Map<String, dynamic>.from(data));
    }
    throw const CloudApiException('分享码返回的数据格式不正确');
  }

  Future<CommunityCharacterGroupInfo> uploadGroup({
    required CloudSession session,
    required CharacterGroupPackage package,
    bool isPublic = true,
  }) async {
    final json = await _requestJson(
      'POST',
      '/api/community/character-groups',
      session: session,
      body: {'is_public': isPublic, 'payload': package.toJson()},
    );
    final data = json['data'];
    if (data is Map) {
      return CommunityCharacterGroupInfo.fromJson(
        Map<String, dynamic>.from(data),
      );
    }
    throw const CloudApiException('上传结果格式不正确');
  }

  Future<CommunityCharacterGroupInfo> updateGroup({
    required CloudSession session,
    required String communityId,
    required CharacterGroupPackage package,
    bool? isPublic,
  }) async {
    final body = <String, dynamic>{'payload': package.toJson()};
    if (isPublic != null) body['is_public'] = isPublic;

    final json = await _requestJson(
      'PUT',
      '/api/community/character-groups/${Uri.encodeComponent(communityId)}',
      session: session,
      body: body,
    );
    final data = json['data'];
    if (data is Map) {
      return CommunityCharacterGroupInfo.fromJson(
        Map<String, dynamic>.from(data),
      );
    }
    throw const CloudApiException('更新结果格式不正确');
  }

  Future<void> deleteGroup({
    required CloudSession session,
    required String communityId,
  }) async {
    await _requestJson(
      'DELETE',
      '/api/community/character-groups/${Uri.encodeComponent(communityId)}',
      session: session,
    );
  }

  Future<Map<String, dynamic>> _requestJson(
    String method,
    String path, {
    Map<String, String>? queryParameters,
    CloudSession? session,
    Map<String, dynamic>? body,
  }) async {
    final uri = ApiConfig.cloudUri(path, queryParameters);
    if (uri == null) {
      throw const CloudApiException('尚未配置云端 API 地址');
    }

    final headers = <String, String>{
      'Accept': 'application/json',
      if (body != null) 'Content-Type': 'application/json',
      if (session != null) 'Authorization': 'Bearer ${session.token}',
    };
    final encodedBody = body == null ? null : jsonEncode(body);

    http.Response response;
    switch (method.toUpperCase()) {
      case 'POST':
        response = await http.post(uri, headers: headers, body: encodedBody);
        break;
      case 'PUT':
        response = await http.put(uri, headers: headers, body: encodedBody);
        break;
      case 'DELETE':
        response = await http.delete(uri, headers: headers, body: encodedBody);
        break;
      case 'GET':
      default:
        response = await http.get(uri, headers: headers);
    }
    final decodedBody = utf8.decode(response.bodyBytes);
    final decoded = decodedBody.isEmpty ? const {} : jsonDecode(decodedBody);
    final root = decoded is Map<String, dynamic>
        ? decoded
        : Map<String, dynamic>.from(decoded as Map);

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw CloudApiException(
        (root['message'] ?? root['error'] ?? '社区请求失败').toString(),
      );
    }
    if (root['status'] != 'success') {
      throw CloudApiException((root['message'] ?? '社区请求失败').toString());
    }
    return root;
  }
}

File: android_app/app/src/main/java/com/animeow/app/data/remote/CharacterGroupCommunityService.kt
package com.animeow.app.data.remote

import com.animeow.app.BuildConfig
import com.animeow.app.data.cloud.CloudSession
import com.animeow.app.data.cloud.cloudRemoteFailure
import java.net.URLEncoder
import org.json.JSONArray
import org.json.JSONObject

data class CommunityCharacterGroup(
    val id: String,
    val name: String,
    val description: String? = null,
    val coverUrl: String? = null,
    val shareCode: String? = null,
    val username: String? = null,
    val characterCount: Int = 0,
    val workCount: Int = 0,
    val downloadCount: Int = 0,
    val isPublic: Boolean = true,
    val updatedAt: String? = null,
)

data class CommunityCharacterGroupPackage(
    val summary: CommunityCharacterGroup,
    val payloadJson: String,
)

internal class CharacterGroupCommunityService(
    private val client: JsonHttpClient = JsonHttpClient(),
) {
    val isConfigured: Boolean get() = BuildConfig.CLOUD_API_BASE.isNotBlank()

    suspend fun list(query: String? = null, limit: Int = 40): List<CommunityCharacterGroup> {
        val keyword = query?.trim().orEmpty()
        val url = buildString {
            append(baseUrl())
            append("/api/community/character-groups?limit=")
            append(limit.coerceIn(1, 80))
            append("&offset=0")
            if (keyword.isNotEmpty()) {
                append("&keyword=")
                append(URLEncoder.encode(keyword, Charsets.UTF_8.name()))
            }
        }
        val root = JSONObject(client.get(url))
        checkSuccess(root)
        val data = root.optJSONObject("data")
        val items = data?.optJSONArray("items") ?: root.optJSONArray("data") ?: JSONArray()
        return buildList {
            for (index in 0 until items.length()) {
                items.optJSONObject(index)?.toCommunitySummary()?.let(::add)
            }
        }
    }

    suspend fun fetch(id: String): CommunityCharacterGroupPackage =
        fetchPath("/api/community/character-groups/${encodePath(id)}")

    suspend fun fetchByShareCode(code: String): CommunityCharacterGroupPackage =
        fetchPath(
            "/api/community/character-groups/share/${encodePath(code.trim().uppercase(java.util.Locale.ROOT))}",
        )

    suspend fun listMine(session: CloudSession): List<CommunityCharacterGroup> = authenticatedRequest {
        val root = JSONObject(client.get(baseUrl() + "/api/user/character-groups", authHeaders(session)))
        checkSuccess(root)
        val items = root.optJSONArray("data") ?: JSONArray()
        buildList {
            for (index in 0 until items.length()) {
                items.optJSONObject(index)?.toCommunitySummary()?.let(::add)
            }
        }
    }

    suspend fun publish(
        session: CloudSession,
        payloadJson: String,
        isPublic: Boolean,
    ): CommunityCharacterGroup = authenticatedRequest {
        val body = JSONObject()
            .put("payload", JSONObject(payloadJson))
            .put("is_public", isPublic)
        val root = JSONObject(client.post(
            baseUrl() + "/api/community/character-groups",
            body.toString(),
            authHeaders(session),
        ))
        checkSuccess(root)
        root.getJSONObject("data").toCommunitySummary() ?: error("发布结果缺少角色组摘要")
    }

    suspend fun update(
        session: CloudSession,
        communityId: String,
        payloadJson: String,
        isPublic: Boolean,
    ): CommunityCharacterGroup = authenticatedRequest {
        val body = JSONObject()
            .put("payload", JSONObject(payloadJson))
            .put("is_public", isPublic)
        val root = JSONObject(client.put(
            baseUrl() + "/api/community/character-groups/${encodePath(communityId)}",
            body.toString(),
            authHeaders(session),
        ))
        checkSuccess(root)
        root.getJSONObject("data").toCommunitySummary() ?: error("更新结果缺少角色组摘要")
    }

    suspend fun delete(session: CloudSession, communityId: String) = authenticatedRequest {
        val root = JSONObject(client.delete(
            baseUrl() + "/api/community/character-groups/${encodePath(communityId)}",
            authHeaders(session),
        ))
        checkSuccess(root)
    }

    private suspend fun fetchPath(path: String): CommunityCharacterGroupPackage {
        val root = JSONObject(client.get(baseUrl() + path))
        checkSuccess(root)
        val data = root.optJSONObject("data") ?: error("社区角色组数据格式不正确")
        val payload = data.optJSONObject("payload") ?: error("社区角色组缺少收藏包")
        return CommunityCharacterGroupPackage(
            summary = data.toCommunitySummary() ?: error("社区角色组摘要缺失"),
            payloadJson = payload.toString(),
        )
    }

    private fun baseUrl(): String = BuildConfig.CLOUD_API_BASE.trim().trimEnd('/')
        .takeIf(String::isNotEmpty)
        ?: error("尚未配置云端 API 地址")

    private fun checkSuccess(root: JSONObject) {
        if (root.optString("status") != "success") {
            error(root.optString("message").ifBlank { "社区请求失败" })
        }
    }

    private fun JSONObject.toCommunitySummary(): CommunityCharacterGroup? {
        val id = optString("community_id").takeIf(String::isNotBlank)
            ?: optString("id").takeIf(String::isNotBlank)
            ?: return null
        return CommunityCharacterGroup(
            id = id,
            name = optString("name").ifBlank { "未命名角色组" },
            description = optString("description").takeIf(String::isNotBlank),
            coverUrl = optString("cover_url").takeIf(String::isNotBlank),
            shareCode = optString("share_code").takeIf(String::isNotBlank),
            username = optString("username").takeIf(String::isNotBlank),
            characterCount = optInt("character_count", 0),
            workCount = optInt("work_count", 0),
            downloadCount = optInt("download_count", 0),
            isPublic = optBoolean("is_public", true),
            updatedAt = optString("updated_at").takeIf(String::isNotBlank),
        )
    }

    private fun authHeaders(session: CloudSession): Map<String, String> = mapOf(
        "Authorization" to "Bearer ${session.token}",
        "Accept" to "application/json",
    )

    private suspend fun <T> authenticatedRequest(block: suspend () -> T): T = try {
        block()
    } catch (exception: RemoteHttpException) {
        throw cloudRemoteFailure(
            statusCode = exception.statusCode,
            responseBody = exception.responseBody,
            authenticated = true,
        )
    }

    private fun encodePath(value: String): String =
        URLEncoder.encode(value, Charsets.UTF_8.name()).replace("+", "%20")
}
